### PR TITLE
[Backport][2.2] Alternative fix for Multi Store Emails issue, Fix Async Emails issues, Fix Multiple Email issues

### PR DESCRIPTION
**Pull request by gwharton**
This PR replaces PR #16461 as for some reason, that PR has become detached from my repo and I don't think it will merge properly.

### Original Pull Request
https://github.com/magento/magento2/pull/18471

### Description
This PR removes the previously introduced transportBuilderByStore class that was introduced in 2.2.4 to fix issue #11740 as this implementation caused unwanted regression/bug when sending multiple emails/async emails.

### Fixed Issues (if relevant)
1. magento/magento2#11740:Sending emails from Admin in Multi-Store Environment defaults to Primary Store
2. magento/magento2#14952:Confirmation emails have no FROM or FROM email address 2.2.4
3. magento/magento2#14945:Store Email Addresses are not used anymore
4. magento/magento2#16355:Magento 2.2.4 not sending from sales sender

### Manual testing scenarios
Scenareo 1
- Single store
- Setup store email addresses
- Complete checkout
- Verify from address in confirmation email

Scenareo 2
- Multiple store
- Setup store email addresses for Store 1
- Setup store email addresses for Store 2
- Complete checkout on Store 1- Verify from address in confirmation email
- Complete checkout on Store 2- Verify from address in confirmation email

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by ihor-sviziev** [comment]
Hi @sidolov,
Any updates from discussing this issue with architects?

**Comment by ihor-sviziev** [comment]
@gwharton is it really backport of some fix from 2.3-develop branch?

**Comment by gwharton** [comment]
The same fault is present in 2.3. Magento is prioritising 2.3 now as i guess 2.3.0 is iminent. There is a new pr for 2.3 and i labelled this one backport. Hopefully they will make the decision on the 2.3 pr and this one will follow without discussion.

**Comment by sidolov** [comment]
Hi @ihor-sviziev , @gwharton , we discussed this case with @buskamuza recently. We still not allowed to add new parameters to public methods, the best way is to deprecate the old method and create the new one, like `setFromByStore` and change all usages to the new method.

**Comment by gwharton** [comment]
ok, someone else will have to pick this one up.

File: lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php [hunk 1]
```diff
@@ -172,12 +172,27 @@ public function setReplyTo($email, $name = null)
 
     /**
      * Set mail from address
+     * 
+     * @deprecated Use setFromByStore
```
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php]: 1. It's better as a comment for `@deprecated` define reason why it was deprecated and use `@see` tag for reference with new method.
2. It's better to use new method with "null" value for $store inside deprecated method to reduce code duplication

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: app/code/Magento/Sales/Model/Order/Email/SenderBuilder.php [hunk 2]
```diff
@@ -28,29 +26,19 @@ class SenderBuilder
...
      * @param Template $templateContainer
      * @param IdentityInterface $identityContainer
      * @param TransportBuilder $transportBuilder
-     * @param TransportBuilderByStore $transportBuilderByStore
      */
     public function __construct(
         Template $templateContainer,
         IdentityInterface $identityContainer,
-        TransportBuilder $transportBuilder,
-        TransportBuilderByStore $transportBuilderByStore = null
```
Comment (ihor-sviziev) [inline app/code/Magento/Sales/Model/Order/Email/SenderBuilder.php]: As part of 2.2 release line this change is not backward compatible. It's better to keep it in constructor and not to use inside class.
As part of 2.3 release line while it's still not released - I think we can do such change

File: lib/internal/Magento/Framework/Mail/Template/TransportBuilderByStore.php [hunk 3]
```diff
@@ -1,55 +0,0 @@
-<?php
```
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/Mail/Template/TransportBuilderByStore.php]: Please revert removing of this class and mark it as `@deprecated` with a comment why it was deprecated and a `@see` tag what we should use instead.

PS: as part of 2.3 release line I think we can remove this class

**Review by ihor-sviziev — DISMISSED** [review]

File: lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php (hunk 1 above)
Comment (gwharton) [reply lib/internal/Magento/Framework/Mail/Template/TransportBuilder.php]: Done

File: app/code/Magento/Sales/Model/Order/Email/SenderBuilder.php (hunk 2 above)
Comment (gwharton) [reply app/code/Magento/Sales/Model/Order/Email/SenderBuilder.php]: Done

File: lib/internal/Magento/Framework/Mail/Template/TransportBuilderByStore.php (hunk 3 above)
Comment (gwharton) [reply lib/internal/Magento/Framework/Mail/Template/TransportBuilderByStore.php]: Done

**Comment by krombel** [comment]
As #18471 (the fix for 2.3) is now merged: Is there any ETA to merge (and release) this fix?

**Comment by gwharton** [comment]
Looks like an alternative solution has already been merged for 2.2. Hence the conflict. See here

https://github.com/magento/magento2/commit/d196dcc8592a40461cfc797b91b694e8c046d969#diff-dfb3ee5a7f863458afea75f5524cd2e5

I haven't tested it.

Options, Leave 2.2 as is, Leave 2.3 as is with alternative fix in place.

or

Update 2.2, reverting the above change and implementing the fix in the same way as 2.2

@ihor-sviziev Any preference on the way forward? I'm tempted to say, if the 2.2 fix already merged is working, then lets just leave 2.2 as is, close this PR and let it diverge.

**Comment by miguelbalparda** [comment]
+1 to let it diverge.

**Comment by krombel** [comment]
I am running 2.2.7 and have this issue => I don't think it is fixed

**Comment by ihor-sviziev** [comment]
Hi @gwharton,
PR to 2.3-develop was finally merged.
Could you fix conflict with 2.2-develop branch?

------
Edit: I didn't saw your comment. Let's ask for architects.


@buskamuza we have following situation:
- in 2.3-develop branch was just merged PR exactly the same as this one. As the same time we have another fix just in 2.2-develop. Both of them fixing the same issue in different ways. I prefer to keep the same realization in 2.3-develop and 2.2-develop branches, so we'll have the same list of issues.

What do you think? Which realization should we keep in 2.2-develop?

**Comment by buskamuza** [comment]
I'd suggest to keep implementations as close to each other as possible, so it'e easier to maintain (agree with @ihor-sviziev ).

A concern about the implementation though: Magento Framework should not know about such term as "store" as it's introduced in a module. Term "scope" should be used ("scopeId" is actually used in the resolver).

**Comment by gwharton** [comment]
ok, leave this with me. I'll take a look over the next week.

**Comment by ihor-sviziev** [comment]
@gwharton ok, please update PR.

**Comment by gwharton** [comment]
Please hold off any changes to this PR until I say it is ready. There are still further commits due.

**Comment by gwharton** [comment]
@magento-engcom-team give me test instance

**Comment by magento-engcom-team** [comment]
Hi @gwharton. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @gwharton, here is your [new Magento instance](https://pr-18472.instances.magento-community.engineering).
Admin access: [https://pr-18472.instances.magento-community.engineering/admin](https://pr-18472.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

**Comment by gwharton** [comment]
@ihor-sviziev I think I'm all done with this one. This is now in line with the 2.3-develop build.

I note the comment about scope and store made by @buskamuza but I'm not sure what to do about it at this stage. The changes made here are the same as the changes already merged into 2.3, which is to use the phrase store. I think it makes more sense if we want to change the wording from store to scope, that this is done in a coordinated manner against both 2.2 and 2.3 at the same time, perhaps in a separate series of PRs.

Hopefully this PR will finally now draw a line under the email from address and name problems that have plagued us for the last 6 months or so. :)

**Comment by ihor-sviziev** [comment]
Hi @gwharton,

I had discussion with @buskamuza, she said that in framework we shouldn't have "store", instead - there should be used "scope" word.

For example senderResolver gets actually not "storeId", but "scopeId":
https://github.com/magento/magento2/blob/2.3-develop/lib/internal/Magento/Framework/Mail/Template/SenderResolverInterface.php#L24

Could you update this PR for rename store to "scope" as part of this PR? Also need to prepare PR for renaming in 2.3- release line

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-3968](https://jira.corp.magento.com/browse/ENGCOM-3968) has been created to process this Pull Request

File: lib/internal/Magento/Framework/Mail/Template/TransportBuilderByStore.php [hunk 4]
```diff
@@ -8,6 +8,13 @@
 
 use Magento\Framework\Mail\MessageInterface;
 
+/**
+ * Class TransportBuilderByStore
+ *
+ * @deprecated The ability to set From address based on store is now available
+ * in the \Magento\Framework\Mail\Template\TransportBuilder class
+ * @see \Magento\Framework\Mail\Template\TransportBuilder::setFromByStore
```
Comment (buskamuza) [inline lib/internal/Magento/Framework/Mail/Template/TransportBuilderByStore.php]: Should be changed to `setFromByScope`

Comment (gwharton) [reply lib/internal/Magento/Framework/Mail/Template/TransportBuilderByStore.php]: Well spotted @buskamuza 

Changes committed.

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-3968](https://jira.corp.magento.com/browse/ENGCOM-3968) has been created to process this Pull Request

**Review by miguelbalparda — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @miguelbalparda, thank you for the review.
[ENGCOM-3968](https://jira.corp.magento.com/browse/ENGCOM-3968) has been created to process this Pull Request

**Comment by ladle3000** [comment]
I have a mix of emails that will use the FROM name and some don't at all. Transactional emails almost always have no FROM name

Whereas built in newsletter module is sending it 90% of the time.

I have tried disabling 3rd party modules, amazon modules, dotmailer module

On 2.2.5 and 2.30 I have the same issue and this fix does nothing for either.

outgoing logs of Mageplaza SMTP module shows this is happening in magento before they leave to the server.  And yes, I have the issue without mageplaza module too.

**Comment by ihor-sviziev** [comment]
Hi @snoroozi,
Try to apply this PR as a patch using https://github.com/cweagans/composer-patches
It should fix your issue

**Comment by ladle3000** [comment]
@ihor-sviziev thank you kindly for the answer.

I presume you just do composer require cweagans/composer-patches from magento root directory.

And then what exactly?

**Comment by ladle3000** [comment]
@gwharton can you give step by steps of how to apply this. I don't have any experience with applying a pr as a patch.

**Comment by crankycyclops** [comment]
@snoroozi and @gwharton, I also need some help with how to apply this. I'm currently developing a module that relies on this being fixed. Thanks!

**EDIT:** Am I correct in assuming that this URL will return a diff that I can apply to my instance? https://patch-diff.githubusercontent.com/raw/magento/magento2/pull/18472.diff

**Comment by gwharton** [comment]
I can't comment on how to apply patches using composer as I don't use that method. There should be plenty of help by Google.

Yes, that is the correct url to obtain a patch.

I run from github and apply patches to my codebase using

```
wget https://patch-diff.githubusercontent.com/raw/magento/magento2/pull/18472.diff
git apply 18472.diff
```
You can get any patch by appending .diff to any pull url in github.

**Comment by ladle3000** [comment]
@gwharton I've got the diff copied to my own site and trying to run from there since we use the vendor directory not app/code for the core files

I get this kind of error 

can't find file to patch at input line 266

Perhaps you used the wrong -p or --strip option?

The text leading up to this was:
--------------------------

|diff --git a/lib/internal/Magento/Framework/Mail/Test/Unit/Template/TransportBuilderTest.php b/lib/internal/Magento/Framework/Mail/Test/Unit/Template/TransportBuilderTest.php
|index 0bbfb37356ca..6c5c48fba1b9 100644
|--- a/lib/internal/Magento/Framework/Mail/Test/Unit/Template/TransportBuilderTest.php
|+++ b/lib/internal/Magento/Framework/Mail/Test/Unit/Template/TransportBuilderTest.php

**Comment by ihor-sviziev** [comment]
Hi @snoroozi,
Here is example how to apply patches: https://twitter.com/s3lf/status/957993636058288128
We're using it, but example with https://github.com/cweagans/composer-patches should be almost the same

**Comment by hostep** [comment]
Just informational, but there is official Magento documentation around how to use the `cweagans/composer-patches` module: https://support.magento.com/hc/en-us/articles/360005484154-Create-a-patch-for-a-Magento-2-Composer-installation-from-a-GitHub-commit

Although it doesn't go over complex PR's like this one which span multiple modules, so you need to split out the diff in multiple diff files, one per module (`magento/module-sales` & `magento/framework`).

**Comment by ladle3000** [comment]
Ok thanks @hostep and @ihor-sviziev I tried for hours yesterday and did manage to figure out splitting to two modules.

Still getting failure on the framework one due to some unexpected ending. I'll have to look more today.

I still don't even know if this PR will fix my issue for sure yet.

**Comment by ladle3000** [comment]
I could not get the framework diff to apply, so I've just copied and pasted all the 6 files from raw to my magento and now get this error on testing an email send

Zend\Mail\Transport\Smtp transport expects either a Sender or at least one From address in the Message; none provided


Really need a fix for this, I'm a merchant running a store and sending out generic looking emails now.

**Comment by gwharton** [comment]
You should only copy the changes, not the entire raw file. By copying the entire raw file, you have effectively overwritten your 2.2.7 files with the 2.2-develop files which may contain other changes not compatible with the rest of 2.2.7.

Just go by the diff, if the line is red (prefixed with a "-"), it is removed, if the line is green (prefixed with a "+"), it is added. The white lines are not changed, and are there for you to locate the lines that have changed.

If you just want to get it to work, you could ignore the files with Test in the filename as they are not needed.

Be aware if this PR doesnt make it into 2.2.8 then when you upgrade to 2.2.8 you will need to make these changes again.

**Comment by ladle3000** [comment]
I'm on 2.3 now. Are these changes still compatible if I go through the
trouble.

**Comment by gwharton** [comment]
PR #18471 is the equivalent fix for 2.3 

This has already been merged into 2.3-develop and will be included in release 2.3.1

You need to revert what you have done, and make the changes according to PR #18471

**Comment by ladle3000** [comment]
So shouldn't copy of raw 2.3 develop files do the trick?

**Comment by gwharton** [comment]
Don't copy raw files from develop into a release version........ full stop. You're asking for trouble as the files may not be compatible.

If you are running 2.3.0 then revert your changes by copying over the files from YOUR release to get you back to the originals. Then if you must apply the PR manually, do so by only making the changes identified in the diff.

**Comment by ladle3000** [comment]
Ok. Good to know. I'm only trying on dev first, but glad to know about not
copying the raw files.

**Comment by qrider** [comment]
I am seeing this for a 2.3.0 site still. From what I can tell 2.3 contains the patch. So not sure why it is not working.

**Comment by gwharton** [comment]
2.3.1 i believe got this. Not 2.3.0.

**Comment by qrider** [comment]
thanks

**Comment by LandonL82** [comment]
Has anyone figured out how to apply this patch to a composer 2.2.7 install?

**Comment by crankycyclops** [comment]
> Has anyone figured out how to apply this patch to a composer 2.2.7 install?

@LandonL82, just make the following string replacements in your text editor before applying the patch:

app/code/Magento/Sales -> vendor/magento/module-sales
lib/internal/Magento/Framework -> vendor/magento/framework

**Comment by sudma** [comment]
@LandonL82 
Here is the official Magento blog post about applying the patches via composer.

https://support.magento.com/hc/en-us/articles/360005484154-Create-a-patch-for-a-Magento-2-Composer-installation-from-a-GitHub-commit

**Comment by tuyennn** [comment]
@gwharton Good to have the backward compatible with TransportBuilder.php on <=2.2.6 on your PR

**Comment by farhadfrz** [comment]
I applied the changes to Magento 2.2.8, but still the admin emails are sent from default email!

**Comment by ladle3000** [comment]
Yeah. I never got this patch working for me unfortunately either. I'm using
an smtp module that the developer did a second check on the send from to
temporarily fix it.

It's the free mage plaza one if you need it. Just open a ticket with them
and they'll add the same for you I'm sure

**Comment by tuyennn** [comment]
On 2.2.6 you should exclude Mail.php from patch because it was not implemented some missing functions. Actually the Mail.php was changed since 2.2.7

**Comment by MiomenteMUC** [comment]
Still got the problem in 2.3.1 for every payment method except Paypal and Amazon Pay.
Expectation:
 - Send BCC Copy to admin when order is placed.
Reality:
 - Customer gets Order Mail, admin does not get Copy.
It´s just working for Paypal and Amazon. Not even the integrated standard Magento prepayment or pick-up at store.
Any solutions? No Error in Logs.
Changes from https://github.com/magento/magento2/pull/18472/files/16afc0962463261e4464456b9bdaad4e8f3be924 are implemented, cause it´s 2.3.1 and was merged in 2.3.0 like mentioned before.

Regards, Loqic

**Comment by ihor-sviziev** [comment]
Hi @Loqic2633,
If you're still having this issue on latest Magento version - please create separate issue for that

Thank you!

**Pull request merged**